### DISCUSSION TITLE
Mamba Installation (xargs: mamba: no such file or directory)

### DISCUSSION THREAD

**Discussion by ehaggard15** [biostars/biostar-handbook-issues#215] [General]
Hello,

I am a relatively new coder and have been trying to install the handbook software on MacOSX (Mojave 10.14.6).  I initially ran the installation of the  handbook (curl http://data.biostarhandbook.com/install.sh | bash, from 11.1) after following the steps in the book (or so I thought), which seemed to go fine but could not get doctor.py to run (command not found).  So I went through the step-by-step installation guide (7 under Installation at https://www.biostarhandbook.com/step-by-step.html#step-by-step).  Everything seemed to be loaded correctly, for instance:

`grep BIOSTAR_HANDBOOK_START ~/.bashrc
`
outputs

`>>> BIOSTAR_HANDBOOK_START`

So the shell customization looks to have been run, and 

mamba seems to have installed: 

**Emmet@Emmets-MacBook-Air ~
```
$ conda install mamba -n base -c conda-forge -y
Collecting package metadata (current_repodata.json): done
Solving environment: done


# All requested packages already installed.
```
but when I try to run the installation script after installing mamba, I get the following:

```
(bioinfo) 
Emmet@Emmets-MacBook-Air ~
$ curl -s http://data.biostarhandbook.com/install/conda.txt | xargs mamba install -q -y
xargs: mamba: No such file or directory
(bioinfo)**

```
It could be something obvious, but I don't seem to quite be having the issues some others have had with conda because everything seems to be installing fine until I try to run the script.  I also tried closing and reopening terminal and running the script in base:

`curl http://data.biostarhandbook.com/install.sh | bash`

but I still get the error:

```
(base) 
Emmet@Emmets-MacBook-Air ~
$ curl http://data.biostarhandbook.com/install.sh | bash
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3068  100  3068    0     0  16491      0 --:--:-- --:--:-- --:--:-- 16583
#
# Bootstrapping the Biostar Handbook ...
#
#
# Installing bioinformatics software
#
xargs: mamba: No such file or directory
(base) 
```

Not clear to me if the find... touch command as described here (https://github.com/biostars/biostar-handbook/discussions/185#discussion-3887488) would be relevant to my issue either (I don't just understand it well enough) so any input is much appreciated.

I also followed the directions in section 11 of the step-by-step installation to set up the doctor.py script.  I ran the command to see if any of the installation worked, and the command runs now, but the output looks like this:

```
Emmet@Emmets-MacBook-Air ~
$ ~/bin/doctor.py
# Doctor! Doctor! Give me the news.
# Checking symptoms ...
# bwa ... Traceback (most recent call last):
  File "/Users/Emmet/bin/doctor.py", line 125, in <module>
    health_check()
  File "/Users/Emmet/bin/doctor.py", line 105, in health_check
    errors += tool_check(tools=TOOLS)
  File "/Users/Emmet/bin/doctor.py", line 69, in tool_check
    res = subprocess.run(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
  File "/Users/Emmet/miniconda3/envs/bioinfo/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/Users/Emmet/miniconda3/envs/bioinfo/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/Users/Emmet/miniconda3/envs/bioinfo/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'bwa'
```
(bioinfo)

**Comment by ljjosephmd** [comment]
I'm having a nearly identical problem. Am trying to set up to go through the BioStars handbook. I have created an 'instance' on the  Google Cloud Console with Ubuntu 20.04.4.  I also get the Doctor's complaints about lines 125, 105, and 69 and the absence of 'bwa'. I get three other line numbers for the other concerns. I am a command line novice so I am clueless about how to fix the errors/problems/concerns reported. I'll slog forward (assuming that is possible) but advice welcome.

**Comment by ialbert** [comment]
sorry about the late reaction, I was traveling last week

make sure to close and open the terminal or source .bashrc

there is nothing special about mamba, it is just a program like any other, and in fact you could replace it with conda (would just operate slower)

**Comment by yaseminsucu** [comment]
I also get the same error, please help??

**Comment by yaseminsucu** [comment]
after installing mamba and opening a new window on the terminal, still gives me an error when I am trying to download the software:

xargs: mamba: no such file or directory...

**Reply by jelebioinfo** [reply]
I've got the same error after folowing the instructions.
xargs: mamba: No such file or directory

**Reply by ialbert** [reply]
https://www.biostarhandbook.com/bashpath.html

and

https://www.biostarhandbook.com/bash-profile.html

**Comment by ialbert** [comment]
The installation instructions indicate a step-by-step troubleshooting to be followed when the automated installation does not proceed as indicated. The link is

https://www.biostarhandbook.com/step-by-step.html

**Reply by yaseminsucu** [reply]
I will look into it, thanks!

**Comment by Estrod** [comment]
Get same problem with `curl http://data.biostarhandbook.com/install.sh | bash` on Ubuntu.

The process did install miniconda and created the environment so I could use `conda activate bioinfo` but had a issue because didn't find mamba.

I went to the step by step installation, used `grep BIOSTAR_HANDBOOK_START ~/.bashrc` and get the `# >>> BIOSTAR_HANDBOOK_START` message, so everything good here.

So, to solve the problem this is what I did:

Jump to step 7 "Install mamba" with the `conda install mamba -n base -c conda-forge -y`
Then jump again to steps 9 and 10, these two steps run without problems.
Next, use `doctor.py` and everything should be OK now.
Finally follow step 12 indications.

Hope this helps in case someone else have this mamba error during installation process

**Reply by ialbert** [reply]
I need to rerun the process myself; it sounds like there might be a slight change as to whether mamba is found within the same shell when run from the command line.

I'll do some tests and see what is up. Technically the `install.sh` performs the exact same commands as the step-by-step page,

But step-by-step process runs in an interactive shell,  the other runs in a subshell, and it seems on some systems there is a difference as to how the bash is set up, most likely how the PATH is updated and which commands get seen.
